src/api/entities/Portfolio/index.ts

```diff
@@ -390,6 +393,143 @@ export abstract class Portfolio extends Entity<UniqueIdentifiers, HumanReadable>
+
+    const settlementsPromise = context.queryMiddlewareV2<EnsuredV2<QueryV2, 'legs'>>(
+      settlementsQuery({
+        identityId,
+        portfolioId,
+        address,
+        ticker,
+      })
+    );
+
+    const portfolioMovementsPromise = context.queryMiddlewareV2<
+      EnsuredV2<QueryV2, 'portfolioMovements'>
+    >(
+      portfolioMovementsQuery({
+        identityId,
+        portfolioId,
+        address,
+        ticker,
+      })
+    );
+
+    const [settlementsResult, portfolioMovementsResult, exists] = await Promise.all([
+      settlementsPromise,
+      portfolioMovementsPromise,
+      this.exists(),
+    ]);
+
+    if (!exists) {
+      throw new PolymeshError({
+        code: ErrorCode.DataUnavailable,
+        message: notExistsMessage,
+      });
+    }
+
+    const data: HistoricSettlement[] = [];
+
+    const portfolioFilter = `${identityId}/${new BigNumber(portfolioId || 0).toString()}`;
+
+    const getDirection = (fromId: string, toId: string): SettlementDirectionEnum => {
+      if (fromId === portfolioFilter) {
+        return SettlementDirectionEnum.Outgoing;
+      }
+      if (toId === portfolioFilter) {
+        return SettlementDirectionEnum.Incoming;
+      }
+      return SettlementDirectionEnum.None;
+    };
+
+    /* eslint-disable @typescript-eslint/no-non-null-assertion */
+    settlementsResult.data.legs.nodes.forEach(node => {
+      const {
+        createdBlock,
+        result: settlementResult,
+        legs: { nodes: legs },
+      } = node!.settlement!;
+
+      const { blockId, hash } = createdBlock!;
+
+      data.push({
+        blockNumber: new BigNumber(blockId),
+        blockHash: hash,
+        status: settlementResult as SettlementResultEnum,
+        accounts: legs[0]!.addresses.map(
+          (accountAddress: string) =>
+            new Account({ address: keyToAddress('0x' + accountAddress, context) }, context)
+        ),
+        legs: legs.map(leg => {
+          const { from, to, fromId, toId, assetId, amount } = leg!;
+          return {
+            asset: new Asset({ ticker: assetId }, context),
+            amount: new BigNumber(amount).shiftedBy(-6),
+            direction: getDirection(fromId, toId),
+            from: middlewareV2PortfolioToPortfolio(from!, context),
+            to: middlewareV2PortfolioToPortfolio(to!, context),
+          };
+        }),
+      });
+    });
+
+    portfolioMovementsResult.data.portfolioMovements.nodes.forEach(node => {
+      const {
+        createdBlock,
+        from,
+        to,
+        fromId,
+        toId,
+        assetId,
+        amount,
+        address: accountAddress,
+      } = node!;
+
+      const { blockId, hash } = createdBlock!;
+
+      data.push({
+        blockNumber: new BigNumber(blockId),
+        blockHash: hash,
+        status: SettlementResultEnum.Executed,
+        accounts: [new Account({ address: keyToAddress('0x' + accountAddress, context) }, context)],
```

Comment:
would be nice to put in a check in `keyToAddress` for the `0x` prefix and add it if needed instead of making the caller know about that detail.